### PR TITLE
Improve the performance of EVP_PKCS82PKEY_ex

### PR DESCRIPTION
**Pull request by mattcaswell**
We can easily find out the keytype which should significantly improve the performance of this function because we don't have to try every loaded decoder.

Partial fix for #20399

### CODE REVIEW

**Review by beldmit — DISMISSED** [review]
LGTM

**Comment by t8m** [comment]
I could easily see this as being a bug - it doesn't make sense to not provide this hint to decoders if we have it almost readily available. So I'd support putting it to 3.0 as well. I assume that would require an OTC exception.

**Review by t8m — APPROVED** [review]

**Comment by beldmit** [comment]
@t8m +1

File: crypto/evp/evp_pkey.c
Comment (levitte) [inline crypto/evp/evp_pkey.c:84]: I would rather see that we use `OBJ_obj2txt`

Comment (mattcaswell) [reply crypto/evp/evp_pkey.c:84]: Why do you see that as better? The advantage of this construction is that it gives you a `const char * `to the name directly. `OBJ_obj2txt` introduces an additional buffer - which seems over the top for this.

Comment (levitte) [reply crypto/evp/evp_pkey.c:84]: Consider the case where a provider provides a key type that we know nothing about, but have included the OID in the OSSL_ALGORITHM name string.  That makes it directly fetchable, even if our OBJ db doesn't have a record for it.

**Comment by mattcaswell** [comment]
Fixup pushed addressing @levitte's comment.

**Review by levitte — APPROVED** [review]

File: crypto/evp/evp_pkey.c
Comment (mattcaswell) [inline crypto/evp/evp_pkey.c:87]: Hmm....actually....does the subject of this comment ever reasonably happen now with this latest change? We should always get an OID out...otherwise it's not PKCS8

Comment (levitte) [reply crypto/evp/evp_pkey.c:87]: Heh, I came back here 'cause I thought of exactly this.  So, drop the comment, I guess?

File: crypto/evp/evp_pkey.c
Comment (t8m) [inline crypto/evp/evp_pkey.c:78]: Do we have any existing constant/macro instead of using a magic number?

**Comment by mattcaswell** [comment]
Fixup pushed addressing latest comments. @levitte, @t8m, @beldmit please take another look.

**Review by t8m — APPROVED** [review]

File: crypto/evp/evp_pkey.c
Comment (beldmit) [inline crypto/evp/evp_pkey.c:82]: Shouldn't we iterate decoders in this case?

Comment (slontis) [reply crypto/evp/evp_pkey.c:82]: It should be falling back in that case I would imagine..

Comment (beldmit) [reply crypto/evp/evp_pkey.c:82]: So we should not return NULL here, right?

Comment (paulidale) [reply crypto/evp/evp_pkey.c:82]: Better might be `keytype = NULL` instead of `return NULL`.  Needing `keytype` to be a pointer not an array which introduces some mess.

Comment (levitte) [reply crypto/evp/evp_pkey.c:82]: > Better might be `keytype = NULL` instead of `return NULL`. Needing `keytype` to be a pointer not an array which introduces some mess.

Just introduce a second variable `char *pkeytype = &keytype`, and pass that.

Comment (mattcaswell) [reply crypto/evp/evp_pkey.c:82]: > Shouldn't we iterate decoders in this case?

Why? Neither of these calls should ever fail in the normal course of events. `OBJ_obj2txt()` will return the OID in textual form if it is something unrecognised. If they do fail then some fundamental error has occurred.

Comment (mattcaswell) [reply crypto/evp/evp_pkey.c:82]: Then we will attempt to match the decoder by OID text string (which all of our own decoders supply). This was the suggestion made by by @levitte in https://github.com/openssl/openssl/pull/20412#discussion_r1121769993 and is the result of using `OBJ_obj2txt` instead of `OBJ_nid2sn` as in the original patch.

Basically we have two choices. Either we use `OBJ_obj2txt` which cannot fail in normal operation, and will either give us the name of the object if it is known, or the OID in textual form if it is unknown. Or we can use the `OBJ_obj2nid(OBJ_nid2sn())` approach (as per the original patch) which will give us the name if it is known, or nothing if it is unknown.

If we use `OBJ_obj2txt` then we *always* have a name of some form except in the case of fundamental error, and so we don't fallback to using a NULL name ever. If we use `OBJ_obj2nid(OBJ_nid2sn())` then we may get a NULL name and we should fallback to the full decoding.

Comment (beldmit) [reply crypto/evp/evp_pkey.c:82]: So if 3rd-party decoder doesn't match an OID string, we consider it as a bug in the decoder, right?

If so, it's worth documenting it.

Comment (levitte) [reply crypto/evp/evp_pkey.c:82]: @beldmit, do you mean a provider that has an OSSL_ALGORITHM item like this:

``` C
    { "blargh", NULL, &blarch_functions },
```

... and then libcrypto gets to handle a PKCS#8 blob with the OID, say, 1.3.6.1.4.1.5168.4711.9999, which we can pretend is the OID for blargh.  How will that get to the correct provider, I guess the question is, right?

There are two ways to handle this:

1.  Add the OID in the name string, i.e. something like this:

    ``` C
        { "blargh:1.3.6.1.4.1.5168.4711.9999", NULL, &blarch_functions },
    ```

    This is the preferred method.

2.  Register the OID in libcrypto's OBJ db, using the upcall number `OSSL_FUNC_CORE_OBJ_CREATE`

    This is less preferred, 'cause it's frankly a lot of extra churn, and it involves dealing with NIDs, something we're trying to get away from.

Comment (beldmit) [reply crypto/evp/evp_pkey.c:82]: @levitte, right, thanks for confirmation! Then I'd say it needs to be documented (at least the 1st way).

Comment (levitte) [reply crypto/evp/evp_pkey.c:82]: Hehe, parallel answering at work here ;-)

Yes, BTW, I would consider it a bug if a provider of algorithms libcrypto knows nothing about fails to provide the OID in some form.  It's not only about PKCS#8, I think that CMS and other similar systems/protocols have similar underlying issues, we just haven't really dealt with them yet.

Comment (levitte) [reply crypto/evp/evp_pkey.c:82]: (just for emphasis: of course it's OK if some algorithm doesn't have an associated OID at all...  it can still be used in all places where an OID isn't expected, and will obviously never appear in structures and objects where an OID is expected)

Comment (levitte) [reply crypto/evp/evp_pkey.c:82]: @beldmit, you might want to read this manual:

https://github.com/openssl/openssl/blob/master/doc/man3/OSSL_ALGORITHM.pod

Do you mean to say that it should be worded more strongly, or would you rather have a note on this in another manual?  Which one, in that case?

Comment (beldmit) [reply crypto/evp/evp_pkey.c:82]: @levitte If I understand correctly, before these changes are merged, we didn't have a mandatory requirement for OID to be present - and now we have it. So adding smth like "OIDs are mandatory in case they are not well-known (present in OID database of OpenSSL) to parse the private keys" looks reasonable here and probably in "provider-decoder" page.

Comment (mattcaswell) [reply crypto/evp/evp_pkey.c:82]: Hmmm. Actually this seems like a change in behaviour, so I'm now wondering whether the original `OBJ_obj2nid(OBJ_nid2sn())` was the correct answer after all...

Comment (levitte) [reply crypto/evp/evp_pkey.c:82]: > Hmmm. Actually this seems like a change in behaviour, so I'm now wondering whether the original `OBJ_obj2nid(OBJ_nid2sn())` was the correct answer after all...

I really don't think so.  However, it's arguable that if `EVP_DECODE_CTX_new_from_pkey()` with the keytype parameter != NULL fails, we could fallback to a second call with the keytype parameter = NULL, on the assumption that the failure of the first call being that it couldn't find a keymgr for the given keytype.

Comment (levitte) [reply crypto/evp/evp_pkey.c:82]: > @levitte If I understand correctly, before these changes are merged, we didn't have a mandatory requirement for OID to be present - and now we have it. So adding smth like `OIDs are mandatory in case they are not well-known (present in OID database of OpenSSL) to parse the private keys` looks reasonable here and probably in "provider-decoder" page.

You forget that this depends on the usage of the algorithm.  If all someone wants to do is `openssl enc -provider blargh -blargh -e < foo.txt`, there's no need for an OID.  So the forcefulness of the message needs to be tempered with regard to usage.

So what I said earlier about this being a bug in the provider was a bit too forceful.  I should have said that *for this usage*, I consider it a bug

Comment (mattcaswell) [reply crypto/evp/evp_pkey.c:82]: > I consider it a bug

Hmm. But, before this change, you could have a decoder that didn't specify the OID in textual form and didn't register the OID anywhere else, and `EVP_PKCS82PKEY_ex` would eventually find and use the decoder (after having tried a long list of other ones first). After this change the decoder will no longer be found.

Comment (levitte) [reply crypto/evp/evp_pkey.c:82]: > > I consider it a bug
> 
> Hmm. But, before this change, you could have a decoder that didn't specify the OID in textual form and didn't register the OID anywhere else, and `EVP_PKCS82PKEY_ex` would eventually find and use the decoder (after having tried a long list of other ones first). After this change the decoder will no longer be found.

Hmmm, on re-reading [`OSSL_ALGORITHM(3)`](https://github.com/openssl/openssl/blob/master/doc/man3/OSSL_ALGORITHM.pod), I'd say it does suggest that for things that involve OIDs, the OID should be present in some form.  It doesn't say out right that things will absolutely fail, but it *is* recommended to, and this is a direct quote:

> make sure that the corresponding **OSSL_ALGORITHM**'s *algorithm_names* include known names as well as the OID itself in canonical decimal text form.  That should cover all scenarios.

To me, this suggests that if used to work with `EVP_PKCS8PKEY_ex()`, it was more by luck than anything else.

But granted, that's a harsh interpretation...

So what do you think about my suggestion of a fallback `EVP_DECODER_CTX_new_from_pkey` call with keytype = NULL?

Comment (levitte) [reply crypto/evp/evp_pkey.c:82]: I'm sorry, I forgot...  the way to detect if no decoders were found is with `OSSL_DECODER_CTX_get_num_decoders()`.  example code that displays this:

``` C
#include <stdio.h>
#include <openssl/decoder.h>
#include <openssl/evp.h>
#include <openssl/err.h>

int main()
{
  EVP_PKEY *key = NULL;
  OSSL_DECODER_CTX *ctx;

  if ((ctx = OSSL_DECODER_CTX_new_for_pkey(&key, NULL, NULL, "blargh",
                                           0, NULL, NULL))) {
    printf("Number of decoders found for 'blargh': %d\n",
           OSSL_DECODER_CTX_get_num_decoders(ctx));
  } else {
    ERR_print_errors_fp(stderr);
  }
  OSSL_DECODER_CTX_free(ctx);
}
```

``` console
$ cc -o foo foo.c -lcrypto
$ ./foo
Number of decoders found for 'blargh': 0
```

Comment (beldmit) [reply crypto/evp/evp_pkey.c:82]: > So what I said earlier about this being a bug in the provider was a bit too forceful. I should have said that _for this usage_, I consider it a bug

Smth like:
```
All the names necessary for using an algorithm (symbolic name for looking up by name, OID for usage in ASN.1 structures etc) should be explicitly provided in case they are not well-known (e.g. present in OID database of OpenSSL)
```

Comment (levitte) [reply crypto/evp/evp_pkey.c:82]: > > So what I said earlier about this being a bug in the provider was a bit too forceful. I should have said that _for this usage_, I consider it a bug
> 
> Smth like: `All the names necessary for using an algorithm (symbolic name for looking up by name, OID for usage in ASN.1 structures etc) should be explicitly provided in case they are not well-known (e.g. present in OID database of OpenSSL)`

PR welcome ;-)

Comment (beldmit) [reply crypto/evp/evp_pkey.c:82]: Sure :)

Comment (t8m) [reply crypto/evp/evp_pkey.c:82]: IMO it is actually wrong to go ahead with the decoding with keytype=NULL. It could be misused to parse incorrect PKCS8 structures where the OID does not match the actual contents of the private key and the decoder would decode it just by chance. Better to fail in that case as currently in this PR. So IMO with the docfix as suggested by @beldmit I think the current PR is good as is.

Comment (mattcaswell) [reply crypto/evp/evp_pkey.c:82]: > It could be misused to parse incorrect PKCS8 structures where the OID does not match the actual contents of the private key and the decoder would decode it just by chance. 

Well that would be a clear bug in the decoder. Conversely failing to register the OID or declare the OID text as an alias is not a *clear* bug (there is at least some debate about whether it is a bug or not) - and we could break existing decoders by requiring it. This would put a big question mark to me over whether or not this is suitable as a bug fix for 3.0, or even whether we would want to introduce it at this late stage for 3.1. Whereas the original version of this PR I would have no such concerns over.

> I'm sorry, I forgot... the way to detect if no decoders were found is with OSSL_DECODER_CTX_get_num_decoders(). example code that displays this:

So your proposal, IIUC, is to keep the logic as it currently is using `OBJ_obj2txt`, but if after the `OSSL_DECODER_CTX_new_for_pkey` we find that `OSSL_DECODER_CTX_get_num_decoders()` is 0, then re-call `OSSL_DECODER_CTX_new_for_pkey` but with a NULL keytype? This sounds plausible.

Comment (levitte) [reply crypto/evp/evp_pkey.c:82]: > So your proposal, IIUC, is to keep the logic as it currently is using `OBJ_obj2txt`, but if after the `OSSL_DECODER_CTX_new_for_pkey` we find that `OSSL_DECODER_CTX_get_num_decoders()` is 0, then re-call `OSSL_DECODER_CTX_new_for_pkey` but with a NULL keytype? This sounds plausible.

Exactly correct.

Comment (beldmit) [reply crypto/evp/evp_pkey.c:82]: See #20422

**Comment by mattcaswell** [comment]
New fixup added as per @levitte's suggestion

**Review by t8m — APPROVED** [review]

File: crypto/evp/evp_pkey.c
Comment (levitte) [inline crypto/evp/evp_pkey.c:94]: But, `dctx == NULL` indicates an error, and I don't think that it's fruitful to retry the decoder path when that happens.  So the check for the decoder fallback should really be `(dctx != NULL && OSSL_DECODER_CTX_get_num_decoders(dctx) == 0)`

**Comment by beldmit** [comment]
Well, how many approvals do we have now? I'd revoke mine...

**Comment by mattcaswell** [comment]
Another fixup addressing @levitte's latest comment. Please take another look

**Review by beldmit — APPROVED** [review]
LGTM

Did you measure the repformance?

**Review by levitte — APPROVED** [review]

File: crypto/evp/evp_pkey.c
Comment (slontis) [inline crypto/evp/evp_pkey.c:80]: I wish we checked p8 == NULL here.. (Maybe off topic).

File: crypto/evp/evp_pkey.c
Comment (slontis) [inline crypto/evp/evp_pkey.c:108]: What is the legacy fallback needed here for?

**Review by slontis — APPROVED** [review]

File: crypto/evp/evp_pkey.c
Comment (levitte) [reply crypto/evp/evp_pkey.c:108]: If there are no decoder because there are no keymgmt implementations, this falls back to try with engines and other legacy mechanisms.

Although, the condition seems a bit odd.  It should probably rather be the same as for the keytype = NULL fallback above, so something like this should probably be done I guess:

``` diff
-    if (dctx == NULL
-        || !OSSL_DECODER_from_data(dctx, &p8_data, &len))
+    if (dctx != NULL && OSSL_DECODER_CTX_get_num_decoders(ctx) == 0)
+        /* Error indicated by |pkey == NULL| */
+        (void)OSSL_DECODER_from_data(dctx, &p8_data, &len);
+    else
         /* try legacy */
```

File: crypto/evp/evp_pkey.c
Comment (mattcaswell) [reply crypto/evp/evp_pkey.c:80]: Actually NOT off topic at all! I just checked. Although master uses the `p8` param without a NULL check, the `i2d_PKCS8_PRIV_KEY_INFO` call *does* check it (eventually....deep in the depths of the ASN1 code) and fails with a 0 return - causing this function to bail out. However, `PKCS8_pkey_get0` will seg fault if passed a NULL `p8` value. So we actually need an explicit check to keep the behaviour the same.

**Comment by mattcaswell** [comment]
New fixup pushed after the comment from @slontis. Sorry....this needs another round of approvals after the latest change: @t8m, @levitte, @beldmit, @slontis.

**Comment by mattcaswell** [comment]
> Did you measure the repformance?

Latest figures. For master:

real	0m15.301s
user	0m15.297s
sys	0m0.004s

For master with this patch:

real	0m8.898s
user	0m8.894s
sys	0m0.004s

**Review by beldmit — APPROVED** [review]
LGTM, nice work!

File: crypto/evp/evp_pkey.c
Comment (levitte) [reply crypto/evp/evp_pkey.c:109]: @mattcaswell, do you consider this relevant?

Comment (mattcaswell) [reply crypto/evp/evp_pkey.c:109]: > do you consider this relevant?

Not within the scope of this PR

Comment (levitte) [reply crypto/evp/evp_pkey.c:109]: Fair enough

**Review by levitte — APPROVED** [review]

**Review by t8m — APPROVED** [review]

**Comment by mattcaswell** [comment]
Pushed to master and 3.1. Thanks for the reviews.

**Comment by mattcaswell** [comment]
**OTC Question: According to the stable release updates policy this PR is not eligible for backport to 3.0. Should we grant an exception for it?**

**Comment by beldmit** [comment]
@mattcaswell my answer will strongly depend on speedup level. Could you please measure it?

**Comment by mattcaswell** [comment]
3.0 without the patch;

real	0m41.978s
user	0m41.977s
sys	0m0.000s

3.0 with the patch:

real	0m12.697s
user	0m12.697s
sys	0m0.000s

**Comment by beldmit** [comment]
Whoops...

**Comment by levitte** [comment]
Oh yeah, that's quite a difference!

**Review by paulidale — APPROVED** [review]

**Comment by paulidale** [comment]
OTC: this should not go into 3.0

**Pull request closed without merging**
